plugins/developer_guide.md

```diff
@@ -0,0 +1,348 @@
+    context.plugin<DokkaBase>().querySingle { samplesTransformer }
+```
+
+You can also browse `DokkaContext` directly, using `single` and `get` methods:
+
+```kotlin
+class SamplePlugin : DokkaPlugin() {
+
+    val extensionPoint by extensionPoint<SampleExtensionPointInterface>()
+    val anotherExtensionPoint by extensionPoint<AnotherSampleExtensionPointInterface>()
+
+    val extension by extending {
+        extensionPoint with SampleExtension()
+    }
+
+    val anotherExtension by extending { 
+        anotherExtensionPoint providing { context ->
+            AnotherSampleExtension(context.single(extensionPoint))
+        }
+    }
+}
+
+interface SampleExtensionPointInterface
+interface AnotherSampleExtensionPointInterface
+
+class SampleExtension: SampleExtensionPointInterface
+class AnotherSampleExtension(sampleExtension: SampleExtensionPointInterface): AnotherSampleExtensionPointInterface
+```
+
+## Dokka Data Model
+
+There a four data models that dokka uses: Documentable Model, Documentation Model, Page Model and Content Model.
+
+### Documentable Model
+
+Documentable model represents parsed data, returned by compiler analysis. It retains basic order structure of parsed `Psi` or `Descriptor` models.
+
+After creation, it is a collection of trees, each with `DModel` as a root. After the Merge step, all trees are folded into one. 
+
+The main building block of this model is `Documentable` class, that is a base class for all more specific types that represents elements of parsed Kotlin and Java classes with pretty self-explanatory names: `DPackage`, `DFunction` and so on. `DClasslike` is a base for class-like elements, such as Classes, Enums, Interfaces and so on.
+
+There are three non-documentable classes important for the model: `DRI`, `SourceSetDependent` and `Extra`.
+
+* `DRI` (Dokka Resource Identifier) is an unique value that identifies specific `Documentable`. All references to other documentables different than direct ownership are described using DRIs. For example, `DFunction` with parameter of type `X` has only X's DRI, not the actual reference to X's Documentable object.
+* `SourceSetDependent` is a map that handles multiplatform data, by connecting platform-specific data, declared with either `expect` or `actual` modifier, to a particular Source Set
+* `ExtraProperty` is used to store any additional information that falls outside of Documentable Model. It is highly recommended to use Extras to provide any additional information when creating own Dokka plugins. To declare a new extra, you need to implement `ExtraProperty` interface:
+```kotlin
+interface ExtraProperty<in C : Any> {
+    interface Key<in C : Any, T : Any> {
+        fun mergeStrategyFor(left: T, right: T): MergeStrategy<C> = MergeStrategy.Fail {
+            throw NotImplementedError("Property merging for $this is not implemented")
+        }
+    }
+
+    val key: Key<C, *>
+}
+```
+It is advised to use following pattern when declaring new extras:
```

Comment:
I think we need here three things:
1. Example of accessing an extra using its class name
2. Explanations of what generic parameter of extra means and how it can be used to narrow down possible targets.
3. Some mention of marker extras (so extras that are objects that are keys to themselves).